### PR TITLE
Fix docker search output form when the description has CJK character

### PR DESCRIPTION
**Pull request by ChiuWang**
**- What I did**
fixes docker search output form when the description has CJK character
**- How I did it**
Every Chinese or Japanese character takes two position of a English charactor, 
But technically many other languages or special symbol should be checked. But I didn't 
find a search result case.
**- How to verify it**
search result
**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**
dog

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
SGTM, anyone can verify current behavior with other writing systems such as Arabic/Hebrew or Mongolian?

**Comment by charrywanganthony** [comment]
As far as i know，any system with usual Chinese(Japanese) fonts will display the character by `East_Asian_Width=Wide` [unicode.org](https://unicode.org/cldr/utility/list-unicodeset.jsp?a=[:East_Asian_Width=Wide:]). Because it's a verification for description so far, i set a general check range`//UTF-8 Kana, CJK, Hangul (from 3000 to DFFF)`.
If it proves valuable, i could make it more accrurate and support more languages later.

**Comment by thaJeztah** [comment]
ping @ripcurld0 you may know some images with other character sets ^^

**Comment by thaJeztah** [comment]
@charrywanganthony looks like you need to `gofmt`;

```
08:23:58 pkg/stringutils/stringutils.go:1::warning: file is not gofmted with -s (gofmt)
08:23:58 pkg/stringutils/stringutils_test.go:1::warning: file is not gofmted with -s (gofmt)
08:23:58 pkg/stringutils/stringutils.go:58:4:warning: should replace displayLen += 1 with displayLen++ (golint)
08:23:58 pkg/stringutils/stringutils.go:66:10:warning: should omit 2nd value from range; this loop is equivalent to `for i := range ...` (golint)
08:23:58 pkg/stringutils/stringutils.go:72:9:warning: should omit 2nd value from range; this loop is equivalent to `for i := range ...` (golint)
08:23:58 pkg/stringutils/stringutils.go:1::warning: file is not goimported (goimports)
08:23:58 pkg/stringutils/stringutils_test.go:1::warning: file is not goimported (goimports)
```

**Comment by AkihiroSuda** [comment]
https://en.m.wikipedia.org/wiki/Halfwidth_and_fullwidth_forms contains some full-width symbols, half-width Katakana and half-width Hangul Jamo

**Comment by ripcurld0** [comment]
@thaJeztah the current code supports Hebrew characters pretty good. No need for code change. However, it is printed in reverse.

```bash
$ docker search ripcurld00dx/boaz
```

**Comment by charrywanganthony** [comment]
I update the commit @thaJeztah , and set a more precise range for  @AkihiroSuda . :)

**Comment by AkihiroSuda** [comment]
Can we use some existing unicode library such as https://godoc.org/golang.org/x/text/width ?

**Comment by charrywanganthony** [comment]
I did the change but met a problem.
```
03:17:03 The result of vndr differs

03:17:03  D vendor/golang.org/x/text/width/gen.go
03:17:03  D vendor/golang.org/x/text/width/gen_common.go
03:17:03  D vendor/golang.org/x/text/width/gen_trieval.go
03:17:03  M vendor/golang.org/x/text/width/kind_string.go
03:17:03  M vendor/golang.org/x/text/width/tables.go
03:17:03  M vendor/golang.org/x/text/width/trieval.go

03:17:03 Please vendor your package with github.com/LK4D4/vndr.
```

When i use vndr i will get the information
`unrecognized import path "golang.org/x/net" (https fetch: Get https://golang.org/x/net?go-get=1: Service Unavailable)` because of Chinese net can't reach golang.org.

How can I change vendor.conf manually to avoid the error

**Comment by AkihiroSuda** [comment]
Please just put `golang.org/x/text/width/gen.go <commit hash>`  to vendor.conf?

File: pkg/stringutils/stringutils.go [hunk 1]
```diff
@@ -33,14 +36,49 @@ func GenerateRandomASCIIString(n int) string {
...
+	}
+
+	var display []int
+	displayLen := 0
+	rs := []rune(s)
+	for _, r := range rs {
+		//In a broad sense, wide characters include W, F, and A, 
+		//see http://unicode.org/reports/tr11/
+		kind := width.LookupRune(r).Kind()
+		if (kind ==1 || kind == 2 || kind == 4) {
```
Comment (AkihiroSuda) [inline pkg/stringutils/stringutils.go]: Please avoid using magic numbers

File: pkg/stringutils/stringutils.go [hunk 2]
```diff
@@ -33,14 +36,52 @@ func GenerateRandomASCIIString(n int) string {
...
+		return string(r[:maxlen-3]) + "..."
+	}
+
+	var display []int
+	displayLen := 0
+	rs := []rune(s)
+	for _, r := range rs {
+		//In a broad sense, wide characters include East Asian Wide, East Asian Fullwidth, and East Asian Ambiguous, 
+		//see http://unicode.org/reports/tr11/
+		EastAsianAmbiguous := 1
```
Comment (AkihiroSuda) [inline pkg/stringutils/stringutils.go]: Therese are already defined as `width.EastAsianAmbiguous`

Comment (charrywanganthony) [reply pkg/stringutils/stringutils.go]: Thx for your help, I correct this, but the vendor problem still exists. @_@

File: vendor.conf [hunk 3]
```diff
@@ -18,6 +18,9 @@ golang.org/x/sys 07c182904dbd53199946ba614a412c61d3c548f5
 github.com/docker/go-units 9e638d38cf6977a37a8ea0078f3ee75a7cdb2dd1
 github.com/docker/go-connections 3ede32e2033de7505e6500d6c868c2b9ed9f169d
 golang.org/x/text f72d8390a633d5dfb0cc84043294db9f6c935756
+golang.org/x/text/width/tables.go dc401f0b44272155af14de5aa0fd4e0ce8e80851
+golang.org/x/text/width/trieval.go fc7fa097411d30e6708badff276c4c164425590c
```
Comment (AkihiroSuda) [inline vendor.conf]: Please remove *.go lines and just update commit hash of `golang.org/x/text`

**Comment by AkihiroSuda** [comment]
看起来不错，除了 `vendor.conf`之外 (LGTM except vendor.conf)

cc @yongtang @coolljt0725 请看一下 (PTAL) ?

File: vendor.conf (hunk 3 above)
Comment (AkihiroSuda) [reply vendor.conf]: > golang.org/x/text/width/gen.go <commit hash>

This was just my copy-paste error. I meant `golang.org/x/text <commit hash>`.
Sorry for causing confusion.

Comment (charrywanganthony) [reply vendor.conf]: problem still exists. @-@ Do I need to update the whole text package?

File: pkg/stringutils/stringutils_test.go [hunk 4]
```diff
@@ -60,17 +60,32 @@ func TestGenerateRandomAsciiStringIsAscii(t *testing.T) {
 func TestEllipsis(t *testing.T) {
 	str := "t🐳ststring"
 	newstr := Ellipsis(str, 3)
-	if newstr != "t🐳s" {
-		t.Fatalf("Expected t🐳s, got %s", newstr)
+	if newstr != "t🐳" {
```
Comment (thaJeztah) [inline pkg/stringutils/stringutils_test.go]: Is this expected? I don't think the whale emoji is a wide character?

Comment (charrywanganthony) [reply pkg/stringutils/stringutils_test.go]: The character do takes two positions of a narrow one. And the test will pass only in this way, after we adopt https://godoc.org/golang.org/x/text/width

Comment (thaJeztah) [reply pkg/stringutils/stringutils_test.go]: Ah! Yes, you're right, I thought it only took a single width, but I see now.

Sorry for the noise 😊

File: vendor.conf (hunk 3 above)
Comment (charrywanganthony) [reply vendor.conf]: Vendor check passed after I update text package. :)

**Comment by thaJeztah** [comment]
WindowsRS1 failure is a known issue, and not related to the changes in this PR; https://github.com/moby/moby/issues/32219

File: pkg/stringutils/stringutils.go [hunk 5]
```diff
@@ -33,14 +36,49 @@ func GenerateRandomASCIIString(n int) string {
...
+	byteLen := len(s)
+	if byteLen == utf8.RuneCountInString(s) {
+		r := []rune(s)
+		if byteLen <= maxlen {
+			return s
+		}
+		if maxlen <= 3 {
+			return string(r[:maxlen])
+		}
+		return string(r[:maxlen-3]) + "..."
```
Comment (thaJeztah) [inline pkg/stringutils/stringutils.go]: If we're improving this function, perhaps this is the right moment to use the actual ellipsis (`"…"`) character for this; not sure if that is one or two positions wide, but at least it allows showing more of the original string 👍

File: pkg/stringutils/stringutils.go [hunk 6]
```diff
@@ -33,14 +36,49 @@ func GenerateRandomASCIIString(n int) string {
...
+	}
+
+	var display []int
+	displayLen := 0
+	rs := []rune(s)
+	for _, r := range rs {
+		//In a broad sense, wide characters include East Asian Wide, East Asian Fullwidth, and East Asian Ambiguous,
+		//see http://unicode.org/reports/tr11/
+		kind := width.LookupRune(r).Kind()
+		if kind == width.EastAsianAmbiguous || kind == width.EastAsianWide || kind == width.EastAsianFullwidth {
```
Comment (thaJeztah) [inline pkg/stringutils/stringutils.go]: Perhaps this should be used to a utility function, something like:

```go
// charWidth returns the number of horizontal positions a character occupies,
// and is used to account for wide characters when displaying strings.
//
// In a broad sense, wide characters include East Asian Wide, East Asian Full-width,
// and East Asian Ambiguous (see http://unicode.org/reports/tr11/).
func charWidth(r rune) int {
	switch width.LookupRune(r).Kind() {
	case width.EastAsianAmbiguous, width.EastAsianWide, width.EastAsianFullwidth:
		return 2
	default:
		return 1
	}
}
```

Then this loop could be something like;

```go
	var (
		display      []int
		displayWidth int
	)
	rs := []rune(s)
	for _, r := range rs {
		cw := charWidth(r)
		displayWidth += cw
		display = append(display, cw)
	}
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
My concern with this PR is that it changes the _behavior_ of the `Ellipsis()`, but not `Truncate()`. Therefore, these functions act different, which is very inconsistent.

For example, the following result in `"t🐳s"`:

```go
stringutils.Truncate("t🐳s", 3)
```

But using _ellipsis_ results in `"t🐳"`:

```go
stringutils.Ellipsis("t🐳s", 3)
```

Looking at this package, I think it has become a collection of not-so-related functionality, and its scope is becoming more vague with this change. Let me try to explain that;
`Ellipsis()` is clearly targeted at _displaying_ strings. `Truncate()` on the other hand is more ambigious; when used for _displaying_ perhaps the string should be truncated at _display_ width, not _number of characters_. However, because the purpose of the package is not well defined, it _can_ be used in other contexts, and those can be problematic (e.g. limit a string's length to a maximum number of characters for storage).

I see this package has also been mentioned in https://github.com/moby/moby/issues/32989 (see [@dnephin's gist](https://gist.github.com/dnephin/35dc10f6b6b7017f058a71908b301d38#file-_-proposal-L167-L170)), and based on the above, I think we should "dismantle" this package, and move code to places it will be used.

Having a look where things are used, and what we can do:

Function | Where it's used | What to do
---|---|---
`GenerateRandomAlphaOnlyString()` | Only used in test (integration-cli) | [integration-cli: `RandomTmpDirPath()`](https://github.com/moby/moby/blob/9be245f438f9fb2eaeb7891673b16aed9262a192/integration-cli/utils_test.go#L55-L68) already mentions that it should not be used. [integration-cli/cli/build/fakestorage/storage.go](https://github.com/moby/moby/blob/9be245f438f9fb2eaeb7891673b16aed9262a192/integration-cli/cli/build/fakestorage/storage.go#L126-L129) uses it. Other uses may not even need a random string; so either move it to the integration-cli, or perhaps it can be removed altogether.
`GenerateRandomASCIIString()` | Looks like it's not used | Probably can be removed
`Ellipsis()` | Only used in the CLI (https://github.com/docker/cli) | Move to that repository, and rewrite to indicate it uses `displayWidth`, not `length`.
`Truncate()` | Looks like it's not used | Move to the, and define its purpose, or perhaps rename it (`TruncateDisplayWidth`, rename the package to `displayutils` and name it `displayutils.TrimWidth()` - bad examples, but just to get you thinking), or remove it altogether
`InSlice()` | used in a couple of tests, and also in the [`seccomp` and `oci_linux` parts](https://github.com/moby/moby/search?utf8=✓&q=stringutils.InSlice&type=) | It's a small function, and given that it's used in different contexts, I think we should copy it to those parts that use it, possibly renaming it for each purpose.
`ShellQuoteArguments()` | Looks like it's not used; also this isn't a "generic" string function - it's specifically for handling _shell_ arguments. | Given that it isn't used, we can probably remove it.

File: pkg/stringutils/stringutils.go (hunk 6 above)
Comment (charrywanganthony) [reply pkg/stringutils/stringutils.go]: OK , I will improve the code after we move the function to CLI repository.

**Comment by charrywanganthony** [comment]
Thanks for your careful work @thaJeztah 
So based on the above, I will
PR[*]: dumplicate the package function by function. 
then
PR: remove the stringutils package in moby
Close this PR
？

File: pkg/stringutils/stringutils.go (hunk 6 above)
Comment (thaJeztah) [reply pkg/stringutils/stringutils.go]: Thanks! Are you interested in working on that move?

Comment (charrywanganthony) [reply pkg/stringutils/stringutils.go]: It will be a great pleasure ：)

**Comment by thaJeztah** [comment]
Awesome! First step would be to open a PR in https://github.com/docker/cli to have these utilities there. 

Once that is done, you can open a new PR in this repository to remove the things we no longer need (and make the other changes)

I suggest we close this PR for now; does that sound good to you?

**Comment by charrywanganthony** [comment]
OK.

**Comment by charrywanganthony** [comment]
But I'm a little confused, what does`open a PR in https://github.com/docker/cli to have these utilities there` exactly mean.

**Comment by thaJeztah** [comment]
That repository contains the Docker CLI; the parts of stringutils that are only used there  (such as `Ellipsis()` should be added to that repository; if you can open a pull request for that.

Happy to give more information/help if you need

**Comment by charrywanganthony** [comment]
Yeah, I know this. What i'm comfused is the stringutils(moby) is in moby/moby/pkg/ ,but stringutils(cli) is in docker/cli/**vendor**/github.com/docker/docker/pkg/, this "vendor-relation" between moby and docker is somewhat equivocal.(sorry I'm a newbee ^-^)

**Comment by thaJeztah** [comment]
Ah yes; what we use in these repositories is "vendoring"; basically, instead of downloading dependencies when compiling, we keep a copy of all dependencies in the repository. 

Doing so, we know _exactly_ which version of the dependency is used (the vendor.conf file contains the exact commit that was used), and also makes sure we can always build (even if the server hosting the dependency is down)

In this case, the Docker CLI uses some packages from this (https://github.com/moby/moby) repository - but still under the old name (https://github.com/docker/docker - that will change soon).


Here's some information about vendoring that I found https://blog.gopheracademy.com/advent-2015/vendor-folder/

**Comment by charrywanganthony** [comment]
If so, IMHO why not just change the code in this repository and then use vendor tools to update the package for CLI. But you recommend `First step would be to open a PR in CLI ... Once that is done...`.

**Comment by thaJeztah** [comment]
Well; the reason for moving it, is that usually, packages in a repository should be in that repository because the repository itself uses it (but they _could_ be useful for other projects)

I this case, the code is not used at all in this repository, so it's better to move it to the repository (Docker CLI) where it's actually used

**Comment by charrywanganthony** [comment]
Thx! sorry for bothering u the beginner level question.

**Comment by thaJeztah** [comment]
Not at all; they're good questions!

**Pull request closed without merging**